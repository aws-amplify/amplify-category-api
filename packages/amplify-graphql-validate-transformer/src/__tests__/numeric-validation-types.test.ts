import { testTransform } from '@aws-amplify/graphql-transformer-test-utils';
import { ModelTransformer } from '@aws-amplify/graphql-model-transformer';
import { ValidateTransformer } from '..';

describe('Numeric Validation Types', () => {
  describe('Invalid usage', () => {
    test.each([
      {
        name: 'rejects empty value',
        schema: /* GraphQL */ `
          type Post @model {
            id: ID!
            rating: Float! @validate(type: gt, value: "")
            score: Int! @validate(type: lt, value: "")
            counts: [Float]! @validate(type: gte, value: "")
            numbers: [Int]! @validate(type: lte, value: "")
          }
        `,
        error: "gt value must be a number. Received '' for field 'rating'",
      },
      {
        name: 'rejects space value',
        schema: /* GraphQL */ `
          type Post @model {
            id: ID!
            rating: Float! @validate(type: gt, value: " ")
            score: Int! @validate(type: lt, value: " ")
            counts: [Float]! @validate(type: gte, value: " ")
            numbers: [Int]! @validate(type: lte, value: " ")
          }
        `,
        error: "lt value must be a number. Received ' ' for field 'rating'",
      },
      {
        name: 'rejects NaN value',
        schema: /* GraphQL */ `
          type Post @model {
            id: ID!
            rating: Float! @validate(type: gt, value: "NaN")
            score: Int! @validate(type: lt, value: "NaN")
            counts: [Float]! @validate(type: gte, value: "NaN")
            numbers: [Int]! @validate(type: lte, value: "NaN")
          }
        `,
        error: "gte value must be a number. Received 'NaN' for field 'rating'",
      },
      {
        name: 'rejects null value',
        schema: /* GraphQL */ `
          type Post @model {
            id: ID!
            rating: Float! @validate(type: gt, value: "null")
            score: Int! @validate(type: lt, value: "null")
            counts: [Float]! @validate(type: gte, value: "null")
            numbers: [Int]! @validate(type: lte, value: "null")
          }
        `,
        error: "lte value must be a number. Received 'null' for field 'rating'",
      },
      {
        name: 'rejects undefined value',
        schema: /* GraphQL */ `
          type Post @model {
            id: ID!
            rating: Float! @validate(type: gt, value: "undefined")
            score: Int! @validate(type: lt, value: "undefined")
            counts: [Float]! @validate(type: gte, value: "undefined")
            numbers: [Int]! @validate(type: lte, value: "undefined")
          }
        `,
        error: "gt value must be a number. Received 'undefined' for field 'rating'",
      },
    ])('$name', ({ schema, error }) => {
      const transformer = new ValidateTransformer();
      expect(() => {
        testTransform({
          schema,
          transformers: [new ModelTransformer(), transformer],
        });
      }).toThrow(error);
    });
  });

  describe('Valid usage', () => {
    test.each([
      {
        name: 'accepts -Infinity value',
        schema: /* GraphQL */ `
          type Post @model {
            id: ID!
            rating: Float! @validate(type: gt, value: "-Infinity")
            score: Int! @validate(type: lt, value: "-Infinity")
            counts: [Float]! @validate(type: gte, value: "-Infinity")
            numbers: [Int]! @validate(type: lte, value: "-Infinity")
          }
        `,
      },
      {
        name: 'accepts Infinity value',
        schema: /* GraphQL */ `
          type Post @model {
            id: ID!
            rating: Float! @validate(type: lt, value: "Infinity")
            score: Int! @validate(type: gt, value: "Infinity")
            counts: [Float]! @validate(type: lte, value: "Infinity")
            numbers: [Int]! @validate(type: gte, value: "Infinity")
          }
        `,
      },
      {
        name: 'accepts extremely large value beyond 64-bit range',
        schema: /* GraphQL */ `
          type Post @model {
            id: ID!
            rating: Float! @validate(type: lte, value: "999999999999999999999999999999")
            score: Int! @validate(type: gte, value: "999999999999999999999999999999")
            counts: [Float]! @validate(type: lte, value: "999999999999999999999999999999")
            numbers: [Int]! @validate(type: gte, value: "999999999999999999999999999999")
          }
        `,
      },
    ])('$name', ({ schema }) => {
      const transformer = new ValidateTransformer();
      expect(() => {
        testTransform({
          schema,
          transformers: [new ModelTransformer(), transformer],
        });
      }).not.toThrow();
    });
  });
});
