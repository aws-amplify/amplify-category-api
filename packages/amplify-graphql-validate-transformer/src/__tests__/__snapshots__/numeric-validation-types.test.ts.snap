// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Numeric Validation Types Valid usage accepts -Infinity value 1`] = `
"type Post {
  id: ID!
  rating: Float!
  score: Int!
  counts: [Float]!
  numbers: [Int]!
  createdAt: AWSDateTime!
  updatedAt: AWSDateTime!
}

input ModelStringInput {
  ne: String
  eq: String
  le: String
  lt: String
  ge: String
  gt: String
  contains: String
  notContains: String
  between: [String]
  beginsWith: String
  attributeExists: Boolean
  attributeType: ModelAttributeTypes
  size: ModelSizeInput
}

input ModelIntInput {
  ne: Int
  eq: Int
  le: Int
  lt: Int
  ge: Int
  gt: Int
  between: [Int]
  attributeExists: Boolean
  attributeType: ModelAttributeTypes
}

input ModelFloatInput {
  ne: Float
  eq: Float
  le: Float
  lt: Float
  ge: Float
  gt: Float
  between: [Float]
  attributeExists: Boolean
  attributeType: ModelAttributeTypes
}

input ModelBooleanInput {
  ne: Boolean
  eq: Boolean
  attributeExists: Boolean
  attributeType: ModelAttributeTypes
}

input ModelIDInput {
  ne: ID
  eq: ID
  le: ID
  lt: ID
  ge: ID
  gt: ID
  contains: ID
  notContains: ID
  between: [ID]
  beginsWith: ID
  attributeExists: Boolean
  attributeType: ModelAttributeTypes
  size: ModelSizeInput
}

input ModelSubscriptionStringInput {
  ne: String
  eq: String
  le: String
  lt: String
  ge: String
  gt: String
  contains: String
  notContains: String
  between: [String]
  beginsWith: String
  in: [String]
  notIn: [String]
}

input ModelSubscriptionIntInput {
  ne: Int
  eq: Int
  le: Int
  lt: Int
  ge: Int
  gt: Int
  between: [Int]
  in: [Int]
  notIn: [Int]
}

input ModelSubscriptionFloatInput {
  ne: Float
  eq: Float
  le: Float
  lt: Float
  ge: Float
  gt: Float
  between: [Float]
  in: [Float]
  notIn: [Float]
}

input ModelSubscriptionBooleanInput {
  ne: Boolean
  eq: Boolean
}

input ModelSubscriptionIDInput {
  ne: ID
  eq: ID
  le: ID
  lt: ID
  ge: ID
  gt: ID
  contains: ID
  notContains: ID
  between: [ID]
  beginsWith: ID
  in: [ID]
  notIn: [ID]
}

enum ModelAttributeTypes {
  binary
  binarySet
  bool
  list
  map
  number
  numberSet
  string
  stringSet
  _null
}

input ModelSizeInput {
  ne: Int
  eq: Int
  le: Int
  lt: Int
  ge: Int
  gt: Int
  between: [Int]
}

enum ModelSortDirection {
  ASC
  DESC
}

type ModelPostConnection {
  items: [Post]!
  nextToken: String
}

input ModelPostFilterInput {
  id: ModelIDInput
  rating: ModelFloatInput
  score: ModelIntInput
  counts: ModelFloatInput
  numbers: ModelIntInput
  createdAt: ModelStringInput
  updatedAt: ModelStringInput
  and: [ModelPostFilterInput]
  or: [ModelPostFilterInput]
  not: ModelPostFilterInput
}

type Query {
  getPost(id: ID!): Post
  listPosts(filter: ModelPostFilterInput, limit: Int, nextToken: String): ModelPostConnection
}

input ModelPostConditionInput {
  rating: ModelFloatInput
  score: ModelIntInput
  counts: ModelFloatInput
  numbers: ModelIntInput
  and: [ModelPostConditionInput]
  or: [ModelPostConditionInput]
  not: ModelPostConditionInput
  createdAt: ModelStringInput
  updatedAt: ModelStringInput
}

input CreatePostInput {
  id: ID
  rating: Float!
  score: Int!
  counts: [Float]!
  numbers: [Int]!
}

input UpdatePostInput {
  id: ID!
  rating: Float
  score: Int
  counts: [Float]
  numbers: [Int]
}

input DeletePostInput {
  id: ID!
}

type Mutation {
  createPost(input: CreatePostInput!, condition: ModelPostConditionInput): Post
  updatePost(input: UpdatePostInput!, condition: ModelPostConditionInput): Post
  deletePost(input: DeletePostInput!, condition: ModelPostConditionInput): Post
}

input ModelSubscriptionPostFilterInput {
  id: ModelSubscriptionIDInput
  rating: ModelSubscriptionFloatInput
  score: ModelSubscriptionIntInput
  counts: ModelSubscriptionFloatInput
  numbers: ModelSubscriptionIntInput
  createdAt: ModelSubscriptionStringInput
  updatedAt: ModelSubscriptionStringInput
  and: [ModelSubscriptionPostFilterInput]
  or: [ModelSubscriptionPostFilterInput]
}

type Subscription {
  onCreatePost(filter: ModelSubscriptionPostFilterInput): Post @aws_subscribe(mutations: [\\"createPost\\"])
  onUpdatePost(filter: ModelSubscriptionPostFilterInput): Post @aws_subscribe(mutations: [\\"updatePost\\"])
  onDeletePost(filter: ModelSubscriptionPostFilterInput): Post @aws_subscribe(mutations: [\\"deletePost\\"])
}
"
`;

exports[`Numeric Validation Types Valid usage accepts Infinity value 1`] = `
"type Post {
  id: ID!
  rating: Float!
  score: Int!
  counts: [Float]!
  numbers: [Int]!
  createdAt: AWSDateTime!
  updatedAt: AWSDateTime!
}

input ModelStringInput {
  ne: String
  eq: String
  le: String
  lt: String
  ge: String
  gt: String
  contains: String
  notContains: String
  between: [String]
  beginsWith: String
  attributeExists: Boolean
  attributeType: ModelAttributeTypes
  size: ModelSizeInput
}

input ModelIntInput {
  ne: Int
  eq: Int
  le: Int
  lt: Int
  ge: Int
  gt: Int
  between: [Int]
  attributeExists: Boolean
  attributeType: ModelAttributeTypes
}

input ModelFloatInput {
  ne: Float
  eq: Float
  le: Float
  lt: Float
  ge: Float
  gt: Float
  between: [Float]
  attributeExists: Boolean
  attributeType: ModelAttributeTypes
}

input ModelBooleanInput {
  ne: Boolean
  eq: Boolean
  attributeExists: Boolean
  attributeType: ModelAttributeTypes
}

input ModelIDInput {
  ne: ID
  eq: ID
  le: ID
  lt: ID
  ge: ID
  gt: ID
  contains: ID
  notContains: ID
  between: [ID]
  beginsWith: ID
  attributeExists: Boolean
  attributeType: ModelAttributeTypes
  size: ModelSizeInput
}

input ModelSubscriptionStringInput {
  ne: String
  eq: String
  le: String
  lt: String
  ge: String
  gt: String
  contains: String
  notContains: String
  between: [String]
  beginsWith: String
  in: [String]
  notIn: [String]
}

input ModelSubscriptionIntInput {
  ne: Int
  eq: Int
  le: Int
  lt: Int
  ge: Int
  gt: Int
  between: [Int]
  in: [Int]
  notIn: [Int]
}

input ModelSubscriptionFloatInput {
  ne: Float
  eq: Float
  le: Float
  lt: Float
  ge: Float
  gt: Float
  between: [Float]
  in: [Float]
  notIn: [Float]
}

input ModelSubscriptionBooleanInput {
  ne: Boolean
  eq: Boolean
}

input ModelSubscriptionIDInput {
  ne: ID
  eq: ID
  le: ID
  lt: ID
  ge: ID
  gt: ID
  contains: ID
  notContains: ID
  between: [ID]
  beginsWith: ID
  in: [ID]
  notIn: [ID]
}

enum ModelAttributeTypes {
  binary
  binarySet
  bool
  list
  map
  number
  numberSet
  string
  stringSet
  _null
}

input ModelSizeInput {
  ne: Int
  eq: Int
  le: Int
  lt: Int
  ge: Int
  gt: Int
  between: [Int]
}

enum ModelSortDirection {
  ASC
  DESC
}

type ModelPostConnection {
  items: [Post]!
  nextToken: String
}

input ModelPostFilterInput {
  id: ModelIDInput
  rating: ModelFloatInput
  score: ModelIntInput
  counts: ModelFloatInput
  numbers: ModelIntInput
  createdAt: ModelStringInput
  updatedAt: ModelStringInput
  and: [ModelPostFilterInput]
  or: [ModelPostFilterInput]
  not: ModelPostFilterInput
}

type Query {
  getPost(id: ID!): Post
  listPosts(filter: ModelPostFilterInput, limit: Int, nextToken: String): ModelPostConnection
}

input ModelPostConditionInput {
  rating: ModelFloatInput
  score: ModelIntInput
  counts: ModelFloatInput
  numbers: ModelIntInput
  and: [ModelPostConditionInput]
  or: [ModelPostConditionInput]
  not: ModelPostConditionInput
  createdAt: ModelStringInput
  updatedAt: ModelStringInput
}

input CreatePostInput {
  id: ID
  rating: Float!
  score: Int!
  counts: [Float]!
  numbers: [Int]!
}

input UpdatePostInput {
  id: ID!
  rating: Float
  score: Int
  counts: [Float]
  numbers: [Int]
}

input DeletePostInput {
  id: ID!
}

type Mutation {
  createPost(input: CreatePostInput!, condition: ModelPostConditionInput): Post
  updatePost(input: UpdatePostInput!, condition: ModelPostConditionInput): Post
  deletePost(input: DeletePostInput!, condition: ModelPostConditionInput): Post
}

input ModelSubscriptionPostFilterInput {
  id: ModelSubscriptionIDInput
  rating: ModelSubscriptionFloatInput
  score: ModelSubscriptionIntInput
  counts: ModelSubscriptionFloatInput
  numbers: ModelSubscriptionIntInput
  createdAt: ModelSubscriptionStringInput
  updatedAt: ModelSubscriptionStringInput
  and: [ModelSubscriptionPostFilterInput]
  or: [ModelSubscriptionPostFilterInput]
}

type Subscription {
  onCreatePost(filter: ModelSubscriptionPostFilterInput): Post @aws_subscribe(mutations: [\\"createPost\\"])
  onUpdatePost(filter: ModelSubscriptionPostFilterInput): Post @aws_subscribe(mutations: [\\"updatePost\\"])
  onDeletePost(filter: ModelSubscriptionPostFilterInput): Post @aws_subscribe(mutations: [\\"deletePost\\"])
}
"
`;

exports[`Numeric Validation Types Valid usage accepts extremely large value beyond 64-bit range 1`] = `
"type Post {
  id: ID!
  rating: Float!
  score: Int!
  counts: [Float]!
  numbers: [Int]!
  createdAt: AWSDateTime!
  updatedAt: AWSDateTime!
}

input ModelStringInput {
  ne: String
  eq: String
  le: String
  lt: String
  ge: String
  gt: String
  contains: String
  notContains: String
  between: [String]
  beginsWith: String
  attributeExists: Boolean
  attributeType: ModelAttributeTypes
  size: ModelSizeInput
}

input ModelIntInput {
  ne: Int
  eq: Int
  le: Int
  lt: Int
  ge: Int
  gt: Int
  between: [Int]
  attributeExists: Boolean
  attributeType: ModelAttributeTypes
}

input ModelFloatInput {
  ne: Float
  eq: Float
  le: Float
  lt: Float
  ge: Float
  gt: Float
  between: [Float]
  attributeExists: Boolean
  attributeType: ModelAttributeTypes
}

input ModelBooleanInput {
  ne: Boolean
  eq: Boolean
  attributeExists: Boolean
  attributeType: ModelAttributeTypes
}

input ModelIDInput {
  ne: ID
  eq: ID
  le: ID
  lt: ID
  ge: ID
  gt: ID
  contains: ID
  notContains: ID
  between: [ID]
  beginsWith: ID
  attributeExists: Boolean
  attributeType: ModelAttributeTypes
  size: ModelSizeInput
}

input ModelSubscriptionStringInput {
  ne: String
  eq: String
  le: String
  lt: String
  ge: String
  gt: String
  contains: String
  notContains: String
  between: [String]
  beginsWith: String
  in: [String]
  notIn: [String]
}

input ModelSubscriptionIntInput {
  ne: Int
  eq: Int
  le: Int
  lt: Int
  ge: Int
  gt: Int
  between: [Int]
  in: [Int]
  notIn: [Int]
}

input ModelSubscriptionFloatInput {
  ne: Float
  eq: Float
  le: Float
  lt: Float
  ge: Float
  gt: Float
  between: [Float]
  in: [Float]
  notIn: [Float]
}

input ModelSubscriptionBooleanInput {
  ne: Boolean
  eq: Boolean
}

input ModelSubscriptionIDInput {
  ne: ID
  eq: ID
  le: ID
  lt: ID
  ge: ID
  gt: ID
  contains: ID
  notContains: ID
  between: [ID]
  beginsWith: ID
  in: [ID]
  notIn: [ID]
}

enum ModelAttributeTypes {
  binary
  binarySet
  bool
  list
  map
  number
  numberSet
  string
  stringSet
  _null
}

input ModelSizeInput {
  ne: Int
  eq: Int
  le: Int
  lt: Int
  ge: Int
  gt: Int
  between: [Int]
}

enum ModelSortDirection {
  ASC
  DESC
}

type ModelPostConnection {
  items: [Post]!
  nextToken: String
}

input ModelPostFilterInput {
  id: ModelIDInput
  rating: ModelFloatInput
  score: ModelIntInput
  counts: ModelFloatInput
  numbers: ModelIntInput
  createdAt: ModelStringInput
  updatedAt: ModelStringInput
  and: [ModelPostFilterInput]
  or: [ModelPostFilterInput]
  not: ModelPostFilterInput
}

type Query {
  getPost(id: ID!): Post
  listPosts(filter: ModelPostFilterInput, limit: Int, nextToken: String): ModelPostConnection
}

input ModelPostConditionInput {
  rating: ModelFloatInput
  score: ModelIntInput
  counts: ModelFloatInput
  numbers: ModelIntInput
  and: [ModelPostConditionInput]
  or: [ModelPostConditionInput]
  not: ModelPostConditionInput
  createdAt: ModelStringInput
  updatedAt: ModelStringInput
}

input CreatePostInput {
  id: ID
  rating: Float!
  score: Int!
  counts: [Float]!
  numbers: [Int]!
}

input UpdatePostInput {
  id: ID!
  rating: Float
  score: Int
  counts: [Float]
  numbers: [Int]
}

input DeletePostInput {
  id: ID!
}

type Mutation {
  createPost(input: CreatePostInput!, condition: ModelPostConditionInput): Post
  updatePost(input: UpdatePostInput!, condition: ModelPostConditionInput): Post
  deletePost(input: DeletePostInput!, condition: ModelPostConditionInput): Post
}

input ModelSubscriptionPostFilterInput {
  id: ModelSubscriptionIDInput
  rating: ModelSubscriptionFloatInput
  score: ModelSubscriptionIntInput
  counts: ModelSubscriptionFloatInput
  numbers: ModelSubscriptionIntInput
  createdAt: ModelSubscriptionStringInput
  updatedAt: ModelSubscriptionStringInput
  and: [ModelSubscriptionPostFilterInput]
  or: [ModelSubscriptionPostFilterInput]
}

type Subscription {
  onCreatePost(filter: ModelSubscriptionPostFilterInput): Post @aws_subscribe(mutations: [\\"createPost\\"])
  onUpdatePost(filter: ModelSubscriptionPostFilterInput): Post @aws_subscribe(mutations: [\\"updatePost\\"])
  onDeletePost(filter: ModelSubscriptionPostFilterInput): Post @aws_subscribe(mutations: [\\"deletePost\\"])
}
"
`;
